### PR TITLE
Add aggregation server client

### PR DESCRIPTION
**Pull request by xichen2020**
cc @cw9 @jeromefroe 

This PR adds the client implementation for writing metrics to the aggregation server. Most of the implementation were moved from `backend/server` package in `m3collector`, with the msgpack encoder replaced with protobuf encoder and hand-written mocks replaced with auto-generated mocks.

### CODE REVIEW

File: client/writer.go [hunk 1]
```diff
@@ -0,0 +1,376 @@
...
+
+			// Otherwise we enqueue the current buffer.
+			buffer := w.prepareEnqueueBufferWithLock(encoder, sizeBefore)
+
+			// Unlock the encoder before we enqueue the old buffer to ensure other
+			// goroutines have an oppurtunity to encode metrics while larger timer
+			// batches are being encoded.
+			encoder.Unlock()
+
+			enqueueErr = w.enqueueBuffer(buffer)
```
Comment (cw9) [inline client/writer.go]: the encoder lock usage here is unconventional and kinda hard to maintain, how much time does this enqueueBuffer call take? is it really necessary to do it outside of lock?

File: client/writer.go [hunk 2]
```diff
@@ -0,0 +1,376 @@
...
+
+			// Re-lock the encoder and update variables since the encoder's buffer
+			// may have been updated.
+			encoder.Lock()
+			sizeBefore = encoder.Len()
+
+			if enqueueErr != nil {
```
Comment (cw9) [inline client/writer.go]: handle enqueueErr here instead of handle it later outside the loop?

File: client/writer.go [hunk 3]
```diff
@@ -0,0 +1,376 @@
...
+				break
+			}
+
+			// If the buffer isn't big enough continue to the next iteration.
+			if sizeAfter := encoder.Len(); sizeAfter < w.flushSize {
+				continue
+			}
+
+			// Otherwise we enqueue the current buffer.
+			buffer := w.prepareEnqueueBufferWithLock(encoder, sizeBefore)
```
Comment (cw9) [inline client/writer.go]: why is this function called for timers specifically than reuse the logic outside the loop?

File: client/writer.go [hunk 4]
```diff
@@ -0,0 +1,376 @@
...
+	for _, encoder := range w.encodersByShard {
+		encoder.Lock()
+		if encoder.Len() == 0 {
+			encoder.Unlock()
+			continue
+		}
+		buffer := encoder.Relinquish()
+		encoder.Unlock()
+		if err := w.enqueueBuffer(buffer); err != nil {
+			multiErr = multiErr.Add(err)
```
Comment (cw9) [inline client/writer.go]: does user handle the err from just one encoder?

File: client/client.go [hunk 5]
```diff
@@ -0,0 +1,233 @@
...
+		if !ok {
+			err = fmt.Errorf("instance %s does not own shard %d", instance.ID(), shardID)
+			multiErr = multiErr.Add(err)
+			continue
+		}
+		if !c.shouldWriteForShard(nowNanos, shard) {
+			continue
+		}
+		if err = c.writerMgr.WriteUntimed(instance, shardID, metric, metadatas); err != nil {
+			multiErr = multiErr.Add(err)
```
[the remark this answers is not in the document]
Comment (xichen2020) [reply client/client.go]: They can log the error and emit metrics to track the errors. Retrying on error is probably not a good idea as the writes are not idempotent. The only time when the error could be surfaced is that the instance queue is full, in which case there probably isn't much you can do.

File: client/conn.go [hunk 6]
```diff
@@ -0,0 +1,225 @@
...
+
+	addr          string
+	connTimeout   time.Duration
+	writeTimeout  time.Duration
+	keepAlive     bool
+	initThreshold int
+	multiplier    int
+	maxThreshold  int
+	maxDuration   time.Duration
+	nowFn         clock.NowFn
```
[the remark this answers is not in the document]
Comment (xichen2020) [reply client/conn.go]: Yeah I can move it down.

File: client/conn.go [hunk 7]
```diff
@@ -0,0 +1,225 @@
...
+	}
+	c.conn = tcpConn
+	return nil
+}
+
+func (c *connection) checkReconnectWithLock() error {
+	// If we haven't accumulated enough failures to warrant another reconnect
+	// and we haven't past the maximum duration since the last time we attempted
+	// to connect then we simply return false without reconnecting.
+	tooManyFailures := c.numFailures > c.threshold
```
[the remark this answers is not in the document]
Comment (xichen2020) [reply client/conn.go]: Yeah see #34 , the main purpose here is to try to recover from a failed server quickly.

File: client/queue.go [hunk 8]
```diff
@@ -0,0 +1,252 @@
...
+
+	"github.com/m3db/m3cluster/placement"
+	"github.com/m3db/m3metrics/encoding/protobuf"
+	"github.com/m3db/m3x/log"
+
+	"github.com/uber-go/tally"
+)
+
+const (
+	defaultQueueSizeNumBuckets = 6
```
[the remark this answers is not in the document]
Comment (xichen2020) [reply client/queue.go]: Sure I think @jeromefroe added this so I can check with him.

File: client/conn.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (cw9) [reply client/conn.go]: what's a typical value for `threshold` and `maxDuration`

File: client/writer.go (hunk 2 above)
Comment (cw9) [reply client/writer.go]: seems `enqueueErr` is only showing up in this code path and `encodeErr` is shared among all the code paths, so it's probably cleaner to handle enqueueErr here and encodeErr outside? but I don't feel strongly either way.

File: client/writer.go (hunk 1 above)
Comment (cw9) [reply client/writer.go]: Sounds good, please make sure there are test cases to guard that neither lock operation could be removed.

File: client/writer.go (hunk 4 above)
Comment (cw9) [reply client/writer.go]: right, but the buffer that failed to be enqueued here will just be discarded I guess?

File: client/writer.go (hunk 3 above)
Comment (cw9) [reply client/writer.go]: right this is still in the loop, make sense

File: client/conn.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (xichen2020) [reply client/conn.go]: The threshold is dynamically adjusted as the number of failures accumulate, the default settings are:
```
	defaultInitReconnectThreshold       = 2
	defaultMaxReconnectThreshold        = 5000
	defaultReconnectThresholdMultiplier = 2
```
Will need to check on the `maxDuration` config.

File: client/writer.go (hunk 1 above)
Comment (xichen2020) [reply client/writer.go]: Yeah this was tested in `TestWriterConcurrentWriteStress` in `writer_test.go`.

File: client/writer.go (hunk 4 above)
Comment (xichen2020) [reply client/writer.go]: Correct, if the queue is full, the buffer will be discarded during the enqueue operation.

**Pull request merged**